packages/rx_bloc_cli/mason_templates/bricks/rx_bloc_base/__brick__/lib/base/routers/router.dart

```diff
@@ -1,17 +1,232 @@
+        _context.read<CoordinatorBlocType>().states.isAuthenticated);
+  }
+
+  final BuildContext _context;
+  late GoRouterRefreshStream refreshListener;
+
+  GoRouter get router => _goRouter;
+
+  late final GoRouter _goRouter = GoRouter(
+    initialLocation: const SplashRoute().location,
+    routes: $appRoutes,
+    redirect: (context, state) async {
+      if (refreshListener.isLoggedIn && state.queryParams['from'] != null) {
+        return state.queryParams['from'];
+      }
+      if (refreshListener.isLoggedIn &&
+          state.subloc == const LoginRoute().location) {
+        return const CounterRoute().location;
+      }
+
+      final pathInfo =
+      router.routeInformationParser.matcher.findMatch(state.location);
+
+      final routeName = RouterPaths.pathToRouteName(pathInfo.fullpath);
+      final hasPermissions =
+      await context.read<PermissionsService>().hasPermission(routeName);
+
+      if (!refreshListener.isLoggedIn && !hasPermissions) {
+        return '${const LoginRoute().location}?from=${state.location}';
+      }
+
+      if (!hasPermissions) {
+        throw AccessDeniedErrorModel();
+      }
+
+      return null;
+    },
+    refreshListenable: GoRouterRefreshStream(
+      _context.read<CoordinatorBlocType>().states.isAuthenticated,
+    ),
+    errorPageBuilder: (context, state) => MaterialPage(
+      key: state.pageKey,
+      child: Scaffold(
+        appBar: AppBar(
+          title: const Text('Error page!'),
+        ),
+        body: const Center(
+          child: Text('Error message.'),
+        ),
+      ),
+    ),
+  );
+}
+
+abstract class RouteData {
+  String get permissionName;
+  String get routeLocation;
+}
+
+@TypedGoRoute<SplashRoute>(path: '/splash')
+@immutable
+class SplashRoute extends GoRouteData implements RouteData {
+  const SplashRoute();
+
+  @override
+  Page<Function> buildPage(BuildContext context, GoRouterState state) =>
+      MaterialPage(
+        key: state.pageKey,
+        child: const SplashPageWithDependencies(),
+      );
+
+  @override
+  String get permissionName => RouterPermissionNames.splashPage;
+
+  @override
+  String get routeLocation => location;
+}
+
+@TypedGoRoute<CounterRoute>(
+  path: '/',
+  routes: <TypedGoRoute<GoRouteData>>[
+    TypedGoRoute<NotificationsRoute>(path: 'notifications'),
+    TypedGoRoute<LoginRoute>(path: 'login'),
+    TypedGoRoute<EnterMessageRoute>(path: 'enterMessage'),
+    TypedGoRoute<ItemsRoute>(path: 'items', routes: <TypedGoRoute<GoRouteData>>[
+      TypedGoRoute<ItemDetailsRoute>(path: ':id'),
+    ]),
   ],
 )
-class $Router {}
+@immutable
+class CounterRoute extends GoRouteData implements RouteData {
+  const CounterRoute();
+
+  @override
+  Page<Function> buildPage(BuildContext context, GoRouterState state) =>
+      MaterialPage(
+        key: state.pageKey,
+        child: const CounterPageWithDependencies(),
+      );
+
+  @override
+  String get permissionName => RouterPermissionNames.counterPage;
```

Comment:
I believe we can rename `RouterPermissionNames` to `PermissionNames`

Reply:
Done